core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -33,12 +33,87 @@ import fs2.internal._
 import Resource.ExitCase
 import Pull._
 
-/** A `p: Pull[F,O,R]` reads values from one or more streams, returns a
-  * result of type `R`, and produces a `Stream[F,O]` when calling `p.stream`.
+/** A pure abstraction representing the intention to perform a sequence of actions
+  * in an effect type `F`, which emits any number of values of an output type `O`.
+  * This sequence may end up in several ways:
+  * a. it may not terminate and keep unrolling new items;
+  * b. it may be aborted by a `Throwable` error;
+  * c. be interrupted by a signal, which may include an error,
+  * d. terminate successfully and returning one single carry-over result of type `R`.
   *
-  * Any resources acquired by `p` are freed following the call to `stream`.
+  * Much like the `IO` values fromo the `cats-effect` library,
+  * `Pull` values are pure, immutable values. Thus, they preserve referential
+  * transparency and are usable in functional programming.
+  * A `Pull` is a data structure that represents just a description
+  * of a sequence of potentially side effectful computation.
   *
-  * Laws:
+  * A pull emits the output values not one-at-a-time, but one-chunk at a time.
+  * A pull only emits non-empty chunks.
+  *
+  * === Pulls variations
+  *
+  * There are some special types of Pull, which are reflected on the arguments of its type parameters:
+  *
+  * A '''pure''' pull is one that performs no effectful actions.
+  * The type of a pure pull has the `F` effect type  bound to `Nothing`.
+  * _Note_ that type `Nothing` is polykinded type, so it can also be
+  * applied as an argument to the type parameter `F[_]`.
+  * Nevertheless, for the sake of clarity, in `fs2` we use the type alias
```

Comment:
We don't necessarily need to mention it here but `Pure` really exists to workaround Scala 2 type inference limitations. When we eventually drop Scala 2, I'd imagine we'll drop `Pure` too.

Reply:
Would it still be helpful, to use `Pure` as an alias of "no effects"?

Reply:
You mean when we get to the point where we can drop Scala 2? I'd rather not but I wouldn't want to break everyone's code over the removal of a type alias. That decision is years and years away though, as I imagine we'll be supporting 2.13 for a very long time.